lib/src/embedded/dispatcher.test.ts

```diff
@@ -0,0 +1,297 @@
+        expect(message.type).toEqual(InboundMessage.MessageCase.IMPORTRESPONSE);
+        done();
+      });
+
+      dispatcher.onImportRequest(() => new InboundMessage.ImportResponse());
+
+      const request = new OutboundMessage.ImportRequest();
+      request.setId(id);
+      outbound$.next({
+        payload: request,
+        type: OutboundMessage.MessageCase.IMPORTREQUEST,
+      });
+    });
+
+    it('triggers the file import request callback', async done => {
+      const id = 1;
+
+      const dispatcher = new Dispatcher(outbound$, message => {
+        const expectedResponse = new InboundMessage.FileImportResponse();
+        expectedResponse.setId(id);
+        expect(message.payload).toEqual(expectedResponse);
+        expect(message.type).toEqual(
+          InboundMessage.MessageCase.FILEIMPORTRESPONSE
+        );
+        done();
+      });
+
+      dispatcher.onFileImportRequest(
+        () => new InboundMessage.FileImportResponse()
+      );
+
+      const request = new OutboundMessage.FileImportRequest();
+      request.setId(id);
+      outbound$.next({
+        payload: request,
+        type: OutboundMessage.MessageCase.FILEIMPORTREQUEST,
+      });
+    });
+
+    it('triggers the canonicalize request callback', async done => {
+      const id = 1;
+
+      const dispatcher = new Dispatcher(outbound$, message => {
+        const expectedResponse = new InboundMessage.CanonicalizeResponse();
+        expectedResponse.setId(id);
+        expect(message.payload).toEqual(expectedResponse);
+        expect(message.type).toEqual(
+          InboundMessage.MessageCase.CANONICALIZERESPONSE
+        );
+        done();
+      });
+
+      dispatcher.onCanonicalizeRequest(
+        () => new InboundMessage.CanonicalizeResponse()
+      );
+
+      const request = new OutboundMessage.CanonicalizeRequest();
+      request.setId(id);
+      outbound$.next({
+        payload: request,
+        type: OutboundMessage.MessageCase.CANONICALIZEREQUEST,
+      });
+    });
+
+    it('triggers the function call request callback', async done => {
+      const id = 1;
+
+      const dispatcher = new Dispatcher(outbound$, message => {
+        const expectedResponse = new InboundMessage.FunctionCallResponse();
+        expectedResponse.setId(id);
+        expect(message.payload).toEqual(expectedResponse);
+        expect(message.type).toEqual(
+          InboundMessage.MessageCase.FUNCTIONCALLRESPONSE
+        );
+        done();
+      });
+
+      dispatcher.onFunctionCallRequest(
+        () => new InboundMessage.FunctionCallResponse()
+      );
+
+      const request = new OutboundMessage.FunctionCallRequest();
+      request.setId(id);
+      outbound$.next({
+        payload: request,
+        type: OutboundMessage.MessageCase.FUNCTIONCALLREQUEST,
+      });
+    });
+  });
+
+  describe('multiple request listeners', () => {
+    it('supports multiple request callbacks', async done => {
```

Comment:
This raises the interesting question of what happens if there are multiple listeners for the same request type. That should probably be an error, since there can only be one response for each request. Alternately, you could statically guarantee that there's only one callback per outbound request by making them constructor parameters (which is close to what we do for the [embedded compiler's `CompileRequest` handler](https://github.com/sass/dart-sass-embedded/blob/e302719e154553d87296291b46b97963e2d18b22/lib/src/dispatcher.dart#L40-L42)).

Reply:
As discussed IRL—these are now passed as constructor params.